cdbbulksearch.py

```diff
@@ -94,80 +189,36 @@ def on_sigint(signal, frame):
         # Linux does not have SIGBREAK.
         pass
 
-    isPGN = args.filename.endswith(".pgn")
-    metalist = []
-    if isPGN:
-        pgn = open(args.filename)
-        while game := chess.pgn.read_game(pgn):
-            metalist.append(game)
-        print(f"Read {len(metalist)} (opening) lines from file {args.filename}.")
-    else:
-        with open(args.filename) as f:
-            for line in f:
-                line = line.strip()
-                if line:
-                    if line.startswith("#"):  # ignore comments
-                        continue
-                    epd, _, moves = line.partition("moves")
-                    epd = " ".join(epd.split()[:4])  # cdb ignores move counters anyway
-                    epdMoves = " moves"
-                    for m in moves.split():
-                        if (
-                            len(m) < 4
-                            or len(m) > 5
-                            or not {m[0], m[2]}.issubset(set("abcdefgh"))
-                            or not {m[1], m[3]}.issubset(set("12345678"))
-                            or (len(m) == 5 and not m[4] in "qrbn")
-                        ):
-                            break
-                        epdMoves += f" {m}"
-                    if epdMoves != " moves":
-                        epd += epdMoves
-                    metalist.append(epd)
-
-    epds = []
-    for item in metalist:
-        if isPGN:
-            epd = item.board().epd()
-            if len(list(item.mainline_moves())):
-                epd += " moves"
-            for move in item.mainline_moves():
-                epd += f" {move}"
-        else:
-            epd = item
-        epds.append(epd)
-
-    print(f"Using {len(epds)} EPDs from file {args.filename}.")
-
-    epdIdx = 0
-    depthLimit = args.depthLimit
-
     executor = concurrent.futures.ProcessPoolExecutor(max_workers=args.bulkConcurrency)
     print(f"Positions to be explored with concurrency {args.bulkConcurrency}.")
 
     tasks = deque()
     task = None
-
-    class TaskCounter:
-        def __init__(self):
-            self.counter = 0
-
-        def inc(self):
-            self.counter += 1
-
-        def dec(self, fn):
-            self.counter -= 1
-
-        def get(self):
-            return self.counter
-
     taskCounter = TaskCounter()
+    first = True
+    epdIdx, epds = 0, []
 
     while True:
         if epdIdx == len(epds):
-            # We arrived at the end of the list: see if we cycle or break.
-            if args.forever:
-                depthLimit += 1
+            # First loop, or we arrived at the end of the list: in that case see if we cycle or break.
+            if first or args.forever:
+                if first or args.reload:
+                    try:
+                        epds = load_epds(args.filename, args.pgnBegin, args.pgnEnd)
+                    except Exception:
+                        if first:
+                            raise Exception(
```

Comment:
I am no expert on exceptions, and do not know what the best way to bail out is here. Feel free to edit and amend.

Reply:
neither am I, but I guess this is fine.